NGitLab/Extensions/FunctionRetryExtensions.cs

```diff
@@ -8,22 +8,32 @@ namespace NGitLab.Extensions
     /// </summary>
     internal static class FunctionRetryExtensions
     {
+        public static Action<string> Logger { get; set; }
+
         /// <summary>
         /// Do a retry a number of time on the received action if it fails
         /// </summary>
-        public static T Retry<T>(this Func<T> action, Func<Exception, int, bool> retryWhen, TimeSpan interval, int retryNumber, bool isIncremental)
+        public static T Retry<T>(this Func<T> action, Func<Exception, int, bool> predicate, TimeSpan waitTime, int maxRetryCount, bool useExponentialBackoff)
         {
-            try
-            {
-                return action();
-            }
-            catch (Exception ex) when (retryWhen(ex, retryNumber))
+            var retriesLeft = maxRetryCount;
+            while (true)
             {
-                Thread.Sleep(interval);
+                try
+                {
+                    return action();
+                }
+                catch (Exception ex) when (predicate(ex, retriesLeft))
```

Comment:
We should force the handling of `retriesLeft`

```suggestion
                catch (Exception ex) when (retriesLeft > 0 && predicate(ex, retriesLeft))
```

Reply:
The `predicate` should handle it, as before. Otherwise what's the point in passing in `retriesLeft` to the predicate?

If a user overrides the default predicate and forgets to take `retriesLeft` into account, well I say tough luck. 😁

Reply:
> Otherwise what's the point in passing in retriesLeft to the predicate?

logging/informative, but I think we should enforce it to avoid bad implementations